src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,105 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final int START_POSITION = 0;
+    private static final double CAPACITY_INCREASE = 1.5;
+    private int size;
+    private T[] elements;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public void ensureCapacity() {
+        if (size == elements.length) {
+            int newCapacity = (int) (size * CAPACITY_INCREASE);
+            T[] newElements = (T[]) new Object[newCapacity];
+            System.arraycopy(elements, START_POSITION, newElements, START_POSITION, size);
+            elements = newElements;
+        }
+    }
+
+    public void indexAddAvailable(int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Can't add value by " + index + " index");
+        }
+    }
+
+    public void indexAvailable(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("element with index " + index + " is not exist");
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity();
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        indexAddAvailable(index);
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexAvailable(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        indexAvailable(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexAvailable(index);
+        T oldValue = elements[index];
+        int numbMove = size - index - 1;
+        if (numbMove > 0) {
+            System.arraycopy(elements, index + 1, elements, index, numbMove);
+        }
+        size--;
+        elements[size] = null;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (elements[i] == null && element == null
+                    || elements[i] != null && elements[i].equals(element)) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("element " + element + " not found");
```

Comment:
The exception message in 'NoSuchElementException' should include more context to be informative. For example, 'Element ' + element + ' not found in the list.'